myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/JHSTaskFactoryImpl.java

```diff
@@ -0,0 +1,290 @@
+        commandInfo.mergeEnvironment(yarnHomeB.build());
+
+        if (myriadExecutorConfiguration.getNodeManagerUri().isPresent()) {
+          //Both FrameworkUser and FrameworkSuperuser to get all of the directory permissions correct.
+          if (!(cfg.getFrameworkUser().isPresent() && cfg.getFrameworkSuperUser().isPresent())) {
+            throw new RuntimeException("Trying to use remote distribution, but frameworkUser" +
+                "and/or frameworkSuperUser not set!");
+          }
+
+          LOGGER.info("Using remote distribution");
+          String clGeneratedCommand = clGenerator.generateCommandLine(profile, null);
+          
+          String nmURIString = myriadExecutorConfiguration.getNodeManagerUri().get();
+
+          //Concatenate all the subcommands
+          String cmd = clGeneratedCommand + " " + executorCmd;
+
+          //get the nodemanagerURI
+          //We're going to extract ourselves, so setExtract is false
+          LOGGER.info("Getting Hadoop distribution from:" + nmURIString);
+          URI nmUri = URI.newBuilder().setValue(nmURIString).setExtract(false)
+              .build();
+
+          //get configs directly from resource manager
+          String configUrlString = clGenerator.getConfigurationUrl();
+          LOGGER.info("Getting config from:" + configUrlString);
+          URI configUri = URI.newBuilder().setValue(configUrlString)
+              .build();
+
+          LOGGER.info("Slave will execute command:" + cmd);
+          commandInfo.addUris(nmUri).addUris(configUri).setValue("echo \"" + cmd + "\";" + cmd);
+          commandInfo.setUser(cfg.getFrameworkSuperUser().get());
+
+        } else {
+          commandInfo.setValue(executorCmd);
+        }
+        return commandInfo.build();
+      }
+      
+      /**
+       * Helper method to reserve ports
+       * @param offer
+       * @param requestedPorts
+       * @return
+       */
+      private List<Long> getAvailablePorts(Offer offer, int requestedPorts) {
+        if (requestedPorts == 0) {
+          return null;
+        }
+        final List<Long> returnedPorts = new ArrayList<>();
+        for (Resource resource : offer.getResourcesList()){
+          if (resource.getName().equals("ports")){
+            Iterator<Value.Range> itr = resource.getRanges().getRangeList().iterator();
+            while (itr.hasNext()) {
+              Value.Range range = itr.next();
+              if (range.getBegin() <= range.getEnd()) {
+                long i = range.getBegin();
+                while (i <= range.getEnd() && returnedPorts.size() < requestedPorts) {
+                  returnedPorts.add(i);
+                  i++;
+                }
+                if (returnedPorts.size() >= requestedPorts) { 
+                  return returnedPorts;
+                }
+              }
+            }
+          }
+        }
+        // this is actually an error condition - we did not have enough ports to use
```

Comment:
It doesn't appear you check the failure case later.  I assume that could would mean not all environment variables are assigned, so we'll go back on YARN defaults, possibly leading to port collisions and eventually a TASK_LOST.

Also, have you you checked the behavior of using mesos assigned ports?  If it's unreliable without a service discovery mechanism it may be something we shouldn't bother with yet.  OTOH, getting Mesos assigned ports to work reliably is the best argument for putting this feature into Myriad.